src/java.base/share/classes/jdk/internal/foreign/abi/ppc64/CallArranger.java

```diff
@@ -0,0 +1,461 @@
+/*
+ * Copyright (c) 2022, 2023, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2023 SAP SE. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.foreign.abi.ppc64;
+
+import jdk.internal.foreign.Utils;
+import jdk.internal.foreign.abi.ABIDescriptor;
+import jdk.internal.foreign.abi.AbstractLinker.UpcallStubFactory;
+import jdk.internal.foreign.abi.Binding;
+import jdk.internal.foreign.abi.CallingSequence;
+import jdk.internal.foreign.abi.CallingSequenceBuilder;
+import jdk.internal.foreign.abi.DowncallLinker;
+import jdk.internal.foreign.abi.LinkerOptions;
+import jdk.internal.foreign.abi.SharedUtils;
+import jdk.internal.foreign.abi.VMStorage;
+import jdk.internal.foreign.abi.ppc64.ABIv2CallArranger;
+
+import java.lang.foreign.AddressLayout;
+import java.lang.foreign.FunctionDescriptor;
+import java.lang.foreign.GroupLayout;
+import java.lang.foreign.MemoryLayout;
+import java.lang.foreign.MemorySegment;
+import java.lang.foreign.ValueLayout;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodType;
+import java.util.List;
+import java.util.Optional;
+
+import static jdk.internal.foreign.abi.ppc64.PPC64Architecture.*;
+import static jdk.internal.foreign.abi.ppc64.PPC64Architecture.Regs.*;
+
+/**
+ * For the PPC64 C ABI specifically, this class uses CallingSequenceBuilder
+ * to translate a C FunctionDescriptor into a CallingSequence, which can then be turned into a MethodHandle.
+ *
+ * This includes taking care of synthetic arguments like pointers to return buffers for 'in-memory' returns.
+ *
+ * There are minor differences between the ABIs implemented on Linux and AIX
+ * which are handled in sub-classes. Clients should access these through the provided
+ * public constants CallArranger.ABIv1/2.
+ */
+public abstract class CallArranger {
+    protected abstract boolean useABIv2();
```

Comment:
This could also be refactored into a static method with the same trick that is used in `LinuxPPC64leLinker::getInstance`. Callers could be static then and you could delete `CallArranger::ABIv2` and `ABIv2CallArranger`.

Reply:
Maybe something like?
```java
    protected static final boolean useABIv2 = CABI.current() == CABI.LINUX_PPC_64_LE;
```

Reply:
That would be better to read, but would make the PPC64 CallArranger dependent on the current CABI.
Note that there are tests which use
```
import jdk.internal.foreign.abi.aarch64.CallArranger;
...
CallArranger.LINUX.getBindings(mt, fd, false);
```
for example. The tests are designed to run on all platforms.

Reply:
> That would be better to read, but would make the PPC64 CallArranger dependent on the current CABI. Note that there are tests which use
> 
> ```
> import jdk.internal.foreign.abi.aarch64.CallArranger;
> ...
> CallArranger.LINUX.getBindings(mt, fd, false);
> ```
> 
> for example. The tests are designed to run on all platforms.

I see, thanks. Would be nice to have some for PPC64 too :)

Reply:
Probably, yes. I didn't find time for figuring out what would be useful tests. We could still add some in the future or with the big endian port.
Another idea: Would the following be better?
`final boolean useABIv2 = (this.getClass() == ABIv2CallArranger.class);`
That would also allow getting rid of the method `useABIv2()`.

Reply:
Or better `final boolean useABIv2 = (this instanceof ABIv2CallArranger);`

Reply:
Please take a look at https://github.com/openjdk/jdk/pull/12708/commits/70736be631e4f1bf3fd3c0d45ddfc076b74ef9dd

Reply:
It looks good.